### PR TITLE
Add byte compile option

### PR DESCRIPTION
**Pull request by bartv**
# Description

This PR adds a bytecompile option to the build command. This option will do two things:

- byte compile all python and remove all python source
- include a dummy so created with cython to force an abi specifc wheel

Closes inmanta/irt#1190

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Comment by bartv** [comment]
Do we need something in the release notes?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Comment by arnaudsjs** [comment]
> Do we need something in the release notes?

I am not sure whether this feature will be useful for end-users, but I also don't see a reason why we wouldn't mention it in the release notes.

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:948]: Could you elaborate on this? It really needs to be `*.pyc`, doesn't it?

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:949]: This could also be "global-exclude *.pyc`, but I don't see how we can reliably handle all such edge cases without overcomplicating this.

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:952]:
```suggestion
                fh.write("\nrecursive-include inmanta_plugins *.pyc\n")
                fh.write("\nglobal-exclude */__pycache__/*\n")
```

File: src/inmanta/moduletool.py
Comment (sanderr) [reply src/inmanta/moduletool.py:949]: Unless we just write it regardless? This is only the builder so nothing we write is actually persisted.

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:961]: Should we document/warn that `setup.py` is overwritten?

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:966]: Isn't `ModuleV2.name` the inmanta module name? I think you might want `ModuleV2Source.get_package_name_for(self_module.name)`.

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:1136]:
```suggestion
        cls, in_folder: str, warn_on_merge: bool = False, build_requires: Optional[list[str]] = None
```
Nice with the built-in `list`!

**Review by sanderr — CHANGES_REQUESTED** [review]
Pretty cool! Couple of concerns about some details.

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:514]: I updated the wording

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:948]: I first checked for *.pyc but, if someone want to limit what is included they will write "something.pyc". In that case we should not add a *.pyc anymore.

So the reasoning is that if there is something with .pyc in there, the module maintainer knows what they are doing.

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:949]: see also the previous comment, then they know what they are doing :-)

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:966]: done

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:1136]: but then we break the api. Should I set the default inside the method then?

File: src/inmanta/moduletool.py
Comment (bartv) [reply src/inmanta/moduletool.py:961]: done

File: src/inmanta/moduletool.py
Comment (sanderr) [inline src/inmanta/moduletool.py:514]:
```suggestion
            help="Produce a module wheel that contains only python bytecode for the plugins.",
```

File: src/inmanta/moduletool.py
Comment (sanderr) [reply src/inmanta/moduletool.py:948]: I see, I guess that makes sense.

File: src/inmanta/moduletool.py
Comment (sanderr) [reply src/inmanta/moduletool.py:966]: I'd like to go back on this. Arnaud just pointed out to me that all this happens in a temporary directory so we don't actually change anything in the project directory.

File: src/inmanta/module.py
Comment (sanderr) [inline src/inmanta/module.py:2388]:
```suggestion
    def _list_python_files(self, plugin_dir: str) -> Iterator[str]:
```
No strong requirement for me but I find the `Iterator` type definition a lot more readable than the `Generator` one. And since this is actually an iterator we could just as well be specific.

File: src/inmanta/plugins.py
Comment (sanderr) [inline src/inmanta/plugins.py:220]:
```suggestion
            line = -1
```
I'm not completely sure, this might break tests, but I believe we already use `-1` in some places where line number does not make sense, e.g. [here](https://github.com/inmanta/inmanta-core/blob/5eb1419d2e142c8f9bb8ba2a28721801c94711dd/src/inmanta/module.py#L1648).

**Review by sanderr — CHANGES_REQUESTED** [review]
Do I understand correctly you would like someone to have a look at this to make compilation and agent code loading work before you merge it?

File: src/inmanta/plugins.py
Comment (bartv) [reply src/inmanta/plugins.py:220]: That does not work. It excepts here: https://github.com/inmanta/inmanta-core/blob/master/src/inmanta/ast/__init__.py#L155

File: src/inmanta/module.py
Comment (bartv) [reply src/inmanta/module.py:2389]: ah, is that correct? did not know that

**Comment by bartv** [comment]
> Do I understand correctly you would like someone to have a look at this to make compilation and agent code loading work before you merge it?

I am still trying myself. However, I have to be realistic on how much time I can invest. So at some point I might need to hand it over to someone else.

File: src/inmanta/loader.py
Comment (wouterdb) [inline src/inmanta/loader.py:175]: I was thinking, would it not be easier to use the magic number to recognize the file. 

It turns out it is now, because the magic changes often. 

https://github.com/python/cpython/blob/main/Lib/importlib/_bootstrap_external.py#L233

Perhaps add a comment here to prevent this same investigation from being done again.

File: src/inmanta/module.py
Comment (wouterdb) [inline src/inmanta/module.py:2410]: This could do with a comment, I think.

File: src/inmanta/module.py
Comment (wouterdb) [inline src/inmanta/module.py:2404]: What if both the pyc and py file are present? 

I think this can happen by accident and should be tested

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/module.py
Comment (bartv) [reply src/inmanta/module.py:2404]: pyc has precedence over py and pyc is invalidated based on timestamp.

Normally pyc files are always created in a __pycache__, so this will only happen when created without our build command.

What else would you like to test?

File: src/inmanta/server/services/codeservice.py
Comment (sanderr) [inline src/inmanta/server/services/codeservice.py:107]: I think this could use a comment. We just don't use the column anymore but we keep it for backwards compatibility? Perhaps we should also add this to the method docstring?

File: src/inmanta/module.py
Comment (sanderr) [reply src/inmanta/module.py:2404]: So suppose a module v1.0.0 is byte-compiled but v2.0.0 isn't (or the other way around), both the compiler and the agent code loader should load the latest version as expected?

File: src/inmanta/loader.py
Comment (sanderr) [inline src/inmanta/loader.py:244]: I'm a bit concerned about this. What if a module contains only one byte compiled file. e.g.
```
inmanta_plugins/mymod/__init__.py  # contains some plugins
inmanta_plugins/mymod/resources.pyc
```
If `install_source` is called in this order, then first `__init__.py` will be written, followed by `resources.pyc`. The loading of the latter will create `__init__.pyc`, which will be an empty file that gets precendence over the actual `__init__.py` file.

Comment (sanderr) [reply src/inmanta/loader.py:244]: We could update this logic but we might just as well consider not doing anything if it's a byte-compiled file, because `touch_inits` is only relevant for orchestrator versions of the modules v1 era anyway. At least, that's how it's documented.

Comment (sanderr) [reply src/inmanta/loader.py:244]: A second remark here is that I don't think an empty `pyc` file is allowed. I get
```
Traceback (most recent call last):
  File "/home/sander/downloads/bytecompiletest/main.py", line 1, in <module>
    import mymod
ImportError: bad magic number in 'mymod': b''
```

So if we would keep this `touch_inits`, we probably need to byte-compile the empty string first.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/agent.py
Comment (bartv) [reply src/inmanta/agent/agent.py:1211]: That is also dangerous because we might overload the server with that

File: src/inmanta/server/services/codeservice.py
Comment (bartv) [reply src/inmanta/server/services/codeservice.py:107]: I keep it here to make sure that we do not accidentally break code or tests. The code that actually uses this, the agent code loading, has been updated.

File: src/inmanta/loader.py
Comment (bartv) [reply src/inmanta/loader.py:246]: Fixed now

File: src/inmanta/agent/agent.py
Comment (wouterdb) [reply src/inmanta/agent/agent.py:1211]: if we don't check if the file already exists on disk, I think this is a step backwards. 

We still download all the files every time (same amount of data*4/3 for the base 64), but we do it in multiple calls (more chatter)

There are two opportunities here:
1. make a raw, binary get_file api (no base64)
2. first check disk, then download. 

We could ticketize these.

File: src/inmanta/module.py
Comment (wouterdb) [reply src/inmanta/module.py:2441]: By where is this filtering done? 

if I look around: 
1. on the exporter code path, I think both the pyc and py files are sent to the server
2. on the loading path, this will cause duplicate loading in 'load_plugins' I think


I also think that if both the py and pyc file are sent over, but the mtime of the files are not also sent, the pyc file will always be recycled. (as per @arnaudsjs research in this matter, iirc, the mtime of the py file is inside the pyc file)

So I would indeed only use the pyc if there is no py file

File: src/inmanta/server/services/codeservice.py
Comment (wouterdb) [reply src/inmanta/server/services/codeservice.py:109]: but it is in fact a breaking change that will break older agents silently? 

I wonder if it would not be better to make it very obvious that this is broken? 

or even make a new API endpoint, and leave this one for backward compatibility and get rid of it in iso6?

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/agent.py
Comment (bartv) [reply src/inmanta/agent/agent.py:1211]: yes, that is what I suggested in the standup at some point

File: src/inmanta/module.py
Comment (sanderr) [reply src/inmanta/module.py:2441]: I'm not completely following. Suppose the yang module is used on an orchestrator somewhere right now. When this is merged we create a byte-compiled artifact for it, that also contains some new functionality. How do we make sure this new functionality will be available if we prefer the `.py` files? There are no `.py` files anymore for the new version, but the ones from the old version are not cleaned up.

Comment (bartv) [reply src/inmanta/module.py:2441]: As it is implemented now, we will select a loader that loads either pyc or py depending on what the server returned. So that means that if v+1 switches to python it will use the python code.

File: src/inmanta/module.py
Comment (bartv) [reply src/inmanta/module.py:2447]: done

File: src/inmanta/module.py
Comment (bartv) [reply src/inmanta/module.py]: done + tests

File: src/inmanta/loader.py
Comment (sanderr) [inline src/inmanta/loader.py:265]: Nice, this does indeed resolve my concerns in this direction. I see how the finder prefers `.py` over `.pyc`, which resolves the concern in the other direction. But wouldn't it be safer to just do the opposite here as well? Then we can never have both files simultaneously, and no assumptions are required between the finder and the loader.

File: src/inmanta/module.py
Comment (sanderr) [inline src/inmanta/module.py:2451]: Any reason for this ordering? It seems inconsistent with the finder's behavior. This way if both files exist the compiler would use the `.py` file but we'd upload the `.pyc` file to the agent.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/loader.py
Comment (sanderr) [reply src/inmanta/loader.py:265]: This comment is not wholly accurate: the finder is not used on the agent iirc, it's just native Python. But I still think it might be best to play safe.

File: src/inmanta/module.py
Comment (sanderr) [reply src/inmanta/module.py:2451]: I don't, that would be fine by me.

File: src/inmanta/agent/agent.py
Comment (wouterdb) [reply src/inmanta/agent/agent.py]: @bartv I completed the nr2, could you link the tickets?

Comment (bartv) [reply src/inmanta/agent/agent.py]: I never created tickets for this. I started working on them to fix them here.

Comment (bartv) [reply src/inmanta/agent/agent.py]: ah no, the binary get api does not have a ticket yet and no work was done on it

File: src/inmanta/server/services/codeservice.py
Comment (wouterdb) [reply src/inmanta/server/services/codeservice.py]: @bartv I changed this to raise a 400 when the source is not valid utf8

File: src/inmanta/data/model.py
Comment (wouterdb) [inline src/inmanta/data/model.py:670]: @bartv I wonder if it would not make more sense to not serve the file_name at all and instead have boolean field 'is_compiled' 

We don't use the file_name anyways.

Comment (bartv) [reply src/inmanta/data/model.py:670]: fine for me

File: src/inmanta/server/services/codeservice.py
Comment (bartv) [reply src/inmanta/server/services/codeservice.py]: that is in the case it is byte compiled?

File: src/inmanta/agent/agent.py
Comment (wouterdb) [reply src/inmanta/agent/agent.py]: https://github.com/inmanta/inmanta-core/issues/4658

File: src/inmanta/server/services/codeservice.py
Comment (wouterdb) [reply src/inmanta/server/services/codeservice.py]: or any other binary file, yes 

I added it to the testcase

File: src/inmanta/data/model.py
Comment (wouterdb) [reply src/inmanta/data/model.py]: done

File: src/inmanta/module.py
Comment (wouterdb) [reply src/inmanta/module.py:2463]: @sanderr @bartv I don't follow this entirely, does this require further action?

Comment (bartv) [reply src/inmanta/module.py:2463]: This was already handled by me

Comment (sanderr) [reply src/inmanta/module.py:2463]: I think so. The issue here is that there is an inconsistency between this method and the finder if both a `.py` and a `.pyc` file are present. Here we prefer the `.pyc` file while the finder does the opposite (unless Bart changed that already, I haven't checked). Bart proposed to keep the behavior here and change the behavior of the finder to match this. I'm on board with that.

Comment (sanderr) [reply src/inmanta/module.py:2463]: Ah, nevermind then.

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/server/services/codeservice.py
Comment (sanderr) [inline src/inmanta/server/services/codeservice.py:109]: This is no longer accurate.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/server/services/codeservice.py
Comment (wouterdb) [reply src/inmanta/server/services/codeservice.py:109]: removed

File: src/inmanta/protocol/endpoints.py
Comment (wouterdb) [reply src/inmanta/protocol/endpoints.py:414]: fixed

File: src/inmanta/loader.py
Comment (wouterdb) [reply src/inmanta/loader.py:270]: fixed

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**
